pkg/export/gcm/promtest/promtest.go

```diff
@@ -0,0 +1,316 @@
+package promtest
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/efficientgo/e2e"
+	"github.com/go-kit/log"
+	"github.com/gogo/protobuf/proto"
+	"github.com/google/go-cmp/cmp"
+	"github.com/oklog/ulid"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/timestamp"
+	"github.com/thanos-io/thanos/pkg/runutil"
+)
+
+// GCMServiceAccountOrFail gets the Google SA JSON content from GCM_SECRET
+// environment variable or fails.
+func GCMServiceAccountOrFail(t testing.TB) []byte {
+	// TODO(bwplotka): Move it to https://cloud.google.com/build CI.
+	saJSON := []byte(os.Getenv("GCM_SECRET"))
+	if len(saJSON) == 0 {
+		t.Fatal("newExportGCM: no GCM_SECRET env var provided, can't run the test")
+	}
+	return saJSON
+}
+
+// ingestionTest represents tests that allows validating Prometheus "backend", so
+// scraper (agent/collector) and DB that serves PromQL e.g.:
+// * export pkg code in this repo and GCM.
+// * GMP collector (built export pkg attached to Prometheus fork) for ingestion and querying.
+// * GMP collector (built export pkg attached to Prometheus fork) and GCM for querying.
+// * Prometheus.
+//
+// This is essentially a simplified compliance tests focused on GCM case and with
+// quick feedback loop (compared to hours with https://github.com/prometheus/compliance/tree/main/promql).
+type ingestionTest struct {
+	t testing.TB
+
+	testID string
+
+	env e2e.Environment
+
+	backends               map[string]backend
+	expectationsPerBackend map[string]model.Matrix
+
+	currTime time.Time
+}
+
+type backend struct {
+	api     v1.API
+	extLset map[string]string
+	b       Backend
+}
+
+// NewIngestionTest takes testing T and returns test instance as well as registry
+// that can be used to register test metrics.
+//
+// NOTE(bwplotka): All test functionality fails tests on error (instead of returning
+// error).
+func NewIngestionTest(t testing.TB, backends []Backend) *ingestionTest {
+	t.Helper()
+
+	// TODO(bwplotka): Consider lazy creation of docker env?
+	e, err := e2e.New()
+	t.Cleanup(e.Close)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	it := &ingestionTest{
+		t: t,
+		// Use ULID as a unique label per test run.
+		// NOTE(bwplotka): Cardinality is obvious, but even for 100 test runs a day with
+		// 100 cases, 10 samples each, that's "only" 10k series / 0.1 million samples a day.
+		testID: ulid.MustNew(ulid.Now(), rand.New(rand.NewSource(time.Now().UnixNano()))).String(),
```

Comment:
Might be easier for debugging purposes to encode the test name here? The `e2e` tests do that to create the namespace for the test.

Reply:
I like that idea, will add 👍🏽